### DISCUSSION TITLE
How to use pvpython with cookiecutter docker scripts

### DISCUSSION THREAD

**Discussion by earlsioson** [Kitware/trame#71] [Q&A] [answered]
Hello,

On our host VM we have an app using VtkRemoteView and "from paraview.web import venv".  We created the app with cookiecutter and it launches fine on the host with the pvpython command.

However, the scripts/build_server.sh fails with error...
```
ModuleNotFoundError: No module named 'paraview'
```

From the paraview-visualizer repo I used some of the env docker settings in the build_server.sh script.
```
docker run -it --rm --gpus all\
    -e TRAME_BUILD_ONLY=1 \
    -e TRAME_PARAVIEW=/opt/paraview   \
    -v "$TRAME_PARAVIEW:/opt/paraview" \
    -v "$DEPLOY_DIR:/deploy" \
    -v "$ROOT_DIR:/local-app"  \
    trame:glvnd
```
I also copied the docker file contents as well...
```
FROM trame:glvnd

# Create the /deploy directory with trame-user as the owner
RUN install -d -o trame-user -g trame-user /deploy

# Install ParaView-EGL
ARG PV_URL=https://www.paraview.org/files/v5.10/ParaView-5.10.0-egl-MPI-Linux-Python3.9-x86_64.tar.gz
RUN mkdir -p /opt/paraview && cd /opt/paraview && wget -qO- $PV_URL | tar --strip-components=1 -xzv
ENV TRAME_PARAVIEW=/opt/paraview

# Copy the server directory
COPY --chown=trame-user:trame-user ./server /deploy/server
```

How can I get the build_server.sh and run_server.sh script to find the paraview module and also execute the app with pvpython instead of regular python?

**Comment by jourdain** [comment]
So to make your app start using pvpython inside your docker you can override the command line that get executed for a given process.

Basically you will do something like below

https://github.com/Kitware/paraview-visualizer/blob/8cfc545972c82d32abb4a08d9b616d8eea203574/docker/setup/apps.yml

**Reply by earlsioson** [reply]
@jourdain, thank you for that link.

How would I get around the error when running build_server.sh...
```
  File "/deploy/server/venv/lib/python3.9/site-packages/pkg_resources/__init__.py", line 489, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/deploy/server/venv/lib/python3.9/site-packages/pkg_resources/__init__.py", line 2852, in load_entry_point
    return ep.load()
  File "/deploy/server/venv/lib/python3.9/site-packages/pkg_resources/__init__.py", line 2443, in load
    return self.resolve()
  File "/deploy/server/venv/lib/python3.9/site-packages/pkg_resources/__init__.py", line 2449, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/deploy/server/venv/lib/python3.9/site-packages/trame_app/__init__.py", line 1, in <module>
    from .app import main
  File "/deploy/server/venv/lib/python3.9/site-packages/trame_app/app/__init__.py", line 1, in <module>
    from .main import main
  File "/deploy/server/venv/lib/python3.9/site-packages/trame_app/app/main.py", line 1, in <module>
    from paraview.web import venv
ModuleNotFoundError: No module named 'paraview'
Build complete. Exiting.
(.venv) earl@dev-trame:~/apps/trame-app/bundles/docker$ 
```

**Comment by jourdain** [comment]
Oh right, you are talking at build time... I'm fixing that problem for trame v2. But for you it is the [`www` generation that fails](https://github.com/Kitware/trame/blob/release-v1/docker/scripts/generate_www.py).

Ideally you want to edit / skip it and manage it yourself outside of that auto build. You will have to run it by hand with pvpython yourself, but as long you place that www dir in the right server spot at the end you will be good to go.

**Comment by jourdain** [comment, marked as answer]
Just wanted to close the loop but that problem should not be an issue anymore with the current trame v2 release and cookie-cutter.

**Reply by earlsioson** [reply]
@jourdain thanks for your suggestion using the paraview-visualizer as reference. I've got this working now with the latest releases.
